DOCS Update optimization docs with NNCF PTQ changes and deprecation of POT

MaximProshin opened the pull request:
### Details:
 - Updated optimization docs with NNCF PTQ changes and deprecation of POT: only structure and content is covered in this PR
 - Updated code snippets (together with @alexsu52)
 - Pictures are updated in **https://github.com/openvinotoolkit/openvino/pull/17421** (by @tsavina)

### Tickets:
 - 94910

Built docs: http://openvino-doc.iotg.sclab.intel.com/optimization-docs/openvino_docs_model_optimization_guide.html

MaximProshin commented:
Built docs: http://openvino-doc.iotg.sclab.intel.com/optimization-docs/openvino_docs_model_optimization_guide.html

tsavina commented on docs/optimization_guide/nncf/ptq/basic_quantization_flow.md:
```diff
@@ -150,12 +149,20 @@ Tune quantization parameters
 See also
 ####################
 
-* `Example of basic quantization flow in PyTorch <https://github.com/openvinotoolkit/nncf/tree/develop/examples/post_training_quantization/torch/mobilenet_v2>`__
```
Maybe we could avoid the repetition of the first part to make it shorter, how about:

Here are some examples of how to apply NNCF post-training quantization to different models:
* `MobileNet v2 OpenVINO Model....
?

tsavina commented on docs/optimization_guide/nncf/ptq/ptq_introduction.md:
```diff
@@ -10,17 +10,17 @@
 
 
-Neural Network Compression Framework (NNCF) provides a new post-training quantization API available in Python that is aimed at reusing the code for model training or validation that is usually available with the model in the source framework, for example, PyTorch or TensroFlow. The API is cross-framework and currently supports models representing in the following frameworks: PyTorch, TensorFlow 2 ...
+Neural Network Compression Framework (NNCF) provides a post-training quantization API available in Python that is aimed at reusing the code for model training or validation that is usually available with the model in the source framework, for example, PyTorch or TensroFlow. The NNCF API is cross-framework and currently supports models in the following frameworks: OpenVINO, PyTorch, TensorFlow 2.x ...
```
```suggestion
Neural Network Compression Framework (NNCF) provides a post-training quantization API available in Python that is aimed at reusing the code for model training or validation that is usually available with the model in the source framework, for example, PyTorch or TensroFlow. The NNCF API is cross-framework and currently supports models in the following frameworks: OpenVINO, PyTorch, TensorFlow 2.x, and ONNX. Currently, post-training quantization for models in OpenVINO Intermediate Representation supports a wide range of methods and has high coverage of various models.
```

MaximProshin replied:
No, it's different meaning after your changes. My idea was to explain that OV IR is better covered than others at tins moment.

tsavina commented on docs/optimization_guide/nncf/ptq/ptq_introduction.md:
```diff
@@ -10,17 +10,17 @@
 
 
-Neural Network Compression Framework (NNCF) provides a new post-training quantization API available in Python that is aimed at reusing the code for model training or validation that is usually available with the model in the source framework, for example, PyTorch or TensroFlow. The API is cross-framework and currently supports models representing in the following frameworks: PyTorch, TensorFlow 2 ...
+Neural Network Compression Framework (NNCF) provides a post-training quantization API available in Python that is aimed at reusing the code for model training or validation that is usually available with the model in the source framework, for example, PyTorch or TensroFlow. The NNCF API is cross-framework and currently supports models in the following frameworks: OpenVINO, PyTorch, TensorFlow 2.x ...
```
most mature is kind of vague

MaximProshin replied:
I agree, but your suggestion changes the meaning.

tsavina commented on docs/optimization_guide/nncf/ptq/quantization_w_accuracy_control.md:
```diff
@@ -7,17 +7,17 @@ Introduction
 This is the advanced quantization flow that allows to apply 8-bit quantization to the model with control of accuracy metric. This is achieved by keeping the most impactful operations within the model in the original precision. The flow is based on the :doc:`Basic 8-bit quantization <basic_quantization_flow>` and has the following differences:
 
-* Beside the calibration dataset, a **validation dataset** is required to compute accuracy metric. They can refer to the same data in the simplest case.
+* Besides the calibration dataset, a **validation dataset** is required to compute accuracy metric. Both datasets can refer to the same data in the simplest case.
```
```suggestion
* Besides the calibration dataset, a **validation dataset** is required to compute the accuracy metric. Both datasets can refer to the same data in the simplest case.
```

tsavina commented on tools/pot/docs/Introduction.md:
```diff
@@ -16,6 +16,8 @@
 
 
 
+.. note:: Post-training Optimization Tool is deprecated since OpenVINO 2023.0. **Neural Network Compression Framework (NNCF)** is recommended for the post-training quantization instead.
```
we might need to add the same note in more POT doc places just in case, like POT FAQ, and probably link to migration tutorial in POT API doc, need to take a look

MaximProshin replied:
I don't really want to add it to every document but feel free to suggest ones where it could be useful. Added note to POT FAQ.

MaximProshin commented on docs/optimization_guide/model_optimization_guide.md:
```diff
@@ -11,34 +11,22 @@
+- :doc:`Training-time Optimization <tmo_introduction>`, a suite of advanced methods for training-time model optimization within the DL framework, such as PyTorch and TensorFlow 2.x. It supports methods, like Quantization-aware Training, Strustured and Unstructured Pruning, etc. 
 
-- :doc:`Training-time Optimization <nncf_ptq_introduction>`, a suite of advanced methods for training-time model optimization within the DL framework, such as PyTorch and TensorFlow 2.x. It supports methods, like Quantization-aware Training and Filter Pruning. NNCF-optimized models can be inferred with OpenVINO using all the available workflows.
+Post-training Quantization is the fastest way to optimize a model and should be applied first, but it is limited in terms of achievable accuracy-performance trade-off. In this case, Training-time Optimization with NNCF is an option.
```
agree

MaximProshin commented on docs/optimization_guide/nncf/ptq/basic_quantization_flow.md:
```diff
@@ -93,14 +92,14 @@ Once the dataset is ready and the model object is instantiated, you can apply 8-
+``nncf.quantize()`` function has several optional parameters that allow tuning the quantization process to get more accurate model. Below is the list of parameters and their description:
 
-* ``model_type`` - used to specify quantization scheme required for specific type of the model. For example, **Transformer** models (BERT, distillBERT, etc.) require a special quantization scheme to preserve accuracy after quantization.
+* ``model_type`` - used to specify quantization scheme required for specific type of the model. ``Transformer`` is only supported special quantization scheme to preserve accuracy after quantization of Transformer models (BERT, distillBERT, etc.). ``None`` is default, i.e. no specific scheme is defined.
```
here I mean that only "Transformer" can be used as "model_type"

AlexKoff88 commented on docs/optimization_guide/model_optimization_guide.md:
```diff
@@ -11,34 +11,22 @@
 
 
-Model optimization is an optional offline step of improving final model performance by applying special optimization methods, such as quantization, pruning, preprocessing optimization, etc. OpenVINO provides several tools to optimize models at different steps of model development:
+Model optimization is an optional offline step of improving the final model performance and reducing the model size by applying special optimization methods, such as 8-bit quantization, pruning, etc. OpenVINO offers two optimization paths implemented in `Neural Network Compression Framework (NNCF) <https://github.com/openvinotoolkit/nncf>`__:
```
Can we also add a note or whatever about the support of the optimized models coming from frameworks? Many times we got a question if the user needs to use our tools for in-framework optimized models.

AlexKoff88 commented:
I am mostly ok with the changes. 

I am also thinking if we need to move "(Experimental) Protecting Model" somewhere (under the POT tab) or mark it.

MaximProshin commented:
> I am mostly ok with the changes.
> 
> I am also thinking if we need to move "(Experimental) Protecting Model" somewhere (under the POT tab) or mark it.

Yes, I also thought about it, so I will move it then.

MaximProshin commented:
@tsavina , @AlexKoff88 , your comments have been applied. Review the updates please.

yury-gorbachev commented on docs/optimization_guide/nncf/ptq/basic_quantization_flow.md:
```diff
@@ -93,14 +92,14 @@ Once the dataset is ready and the model object is instantiated, you can apply 8-
 .. note:: The ``model`` is an instance of the ``torch.nn.Module`` class for PyTorch, ``onnx.ModelProto`` for ONNX, and ``openvino.runtime.Model`` for OpenVINO.
 
-After that the model can be exported into th OpenVINO Intermediate Representation if needed and run faster with OpenVINO.
+After that the model can be exported into the OpenVINO Intermediate Representation if needed and run faster with OpenVINO.
```
I think we need to have section that shows how to do it. It is not very intuitive.
Moreover, there is description of what model is, but no description of what optimized_model is. Is it OV model? Can I compile it right away?

MaximProshin replied:
I've added code example. Please check.

yury-gorbachev commented on docs/optimization_guide/nncf/ptq/quantization_w_accuracy_control.md:
```diff
@@ -51,14 +51,24 @@ Now, you can run quantization with accuracy control. The following code snippet
+* ``drop_type`` defines how the accuracy drop will be calculated: ``ABSOLUTE`` (used by default) or ``RELATIVE``.
+
+``nncf.quantize_with_accuracy_control()`` API supports all the parameters from :doc:`Basic 8-bit quantization <basic_quantization_flow>` API and you can also use them to quantize a model with accuracy control and a custom configuration.
+
```
I think it is worth adding what to do with model later. You already have statement in basic flow that model could be serialized (I have comment for that section also - example missing there).

MaximProshin replied:
I've added code example and some text. Please check it.

yury-gorbachev commented:
General comment is that I recommend to read through other documents as well, sometimes there are misspellings and missing letters, etc.

MaximProshin commented on docs/optimization_guide/model_optimization_guide.md:
```diff
@@ -8,37 +8,24 @@
+- :doc:`Training-time Optimization <tmo_introduction>`, a suite of advanced methods for training-time model optimization within the DL framework, such as PyTorch and TensorFlow 2.x. It supports methods like Quantization-aware Training, Structured and Unstructured Pruning, etc. 
 
-- :doc:`Training-time Optimization <nncf_ptq_introduction>`, a suite of advanced methods for training-time model optimization within the DL framework, such as PyTorch and TensorFlow 2.x. It supports methods, like Quantization-aware Training and Filter Pruning. NNCF-optimized models can be inferred with OpenVINO using all the available workflows.
+.. note:: OpenVINO also supports optimized models (for example, quantized) from source frameworks such as PyTorch, TensorFlow, and ONNX (in Q/DQ format).
```
added

MaximProshin commented on docs/optimization_guide/model_optimization_guide.md:
```diff
@@ -8,37 +8,24 @@
-- :doc:`Training-time Optimization <nncf_ptq_introduction>`, a suite of advanced methods for training-time model optimization within the DL framework, such as PyTorch and TensorFlow 2.x. It supports methods, like Quantization-aware Training and Filter Pruning. NNCF-optimized models can be inferred with OpenVINO using all the available workflows.
+.. note:: OpenVINO also supports optimized models (for example, quantized) from source frameworks such as PyTorch, TensorFlow, and ONNX (in Q/DQ format).
 
+Post-training Quantization is the fastest way to optimize a model and should be applied first, but it is limited in terms of achievable accuracy-performance trade-off. In this case, Training-time Optimization is an option.
```
updated the sentence

MaximProshin commented on docs/optimization_guide/model_optimization_guide.md:
```diff
@@ -8,37 +8,24 @@
-
-Once the model is optimized using the aforementioned tools it can be used for inference using the regular OpenVINO inference workflow. No changes to the inference code are required.
-
 .. image:: _static/images/WHAT_TO_USE.svg
```
pictures were updated in https://github.com/openvinotoolkit/openvino/pull/17421 but I believe it doesn't address your comment. let's discuss how to change these pictures.

MaximProshin commented on docs/optimization_guide/model_optimization_guide.md:
```diff
@@ -8,37 +8,24 @@
-If you are not familiar with model optimization methods, refer to :doc:`post-training methods <pot_introduction>`.
-
 Additional Resources
 ####################
```
added the refs

MaximProshin commented on docs/optimization_guide/nncf/ptq/ptq_introduction.md:
```diff
@@ -10,17 +10,17 @@
-* :doc:`Basic quantization <basic_quantization_flow>` - the simplest quantization flow that allows to apply 8-bit integer quantization to the model.
-* :doc:`Quantization with accuracy control <quantization_w_accuracy_control>` - the most advanced quantization flow that allows to apply 8-bit quantization to the model with accuracy control.
+* :doc:`Basic quantization <basic_quantization_flow>` - the simplest quantization flow that allows applying 8-bit integer quantization to the model.
+* :doc:`Quantization with accuracy control <quantization_w_accuracy_control>` - the most advanced quantization flow that allows applying 8-bit quantization to the model with accuracy control.
```
added to both

alexsu52 commented on docs/optimization_guide/nncf/ptq/basic_quantization_flow.md:
```diff
@@ -150,12 +149,20 @@ Tune quantization parameters
 
+* ``target_device`` - defines the target device, the specificity of which will be taken into account during optimization. The following values are supported: ``ANY`` (default), ``CPU``, ``CPU_SPR``, ``GPU``, and ``VPU``.
+
+* ``advanced_parameters`` - used to specify advanced quantization parameters for fine-tuning the quantization algorithm. Defined by ``nncf.AdvancedQuantizationParameters`` class. ``None`` is default.
```
What needs to be added here?

helena-intel replied:
I'm wondering if there is a reference to AdvancedQuantizationParamaters? It feels like something is missing, the docs say we can add this, that None is the default, but not what the possible values are.

MaximProshin replied:
Right, we discussed internally whether we should describe it on OV level and decided that it's too advanced. I plan to add a cross-reference to NNCF class description from a new NNCF API documentation.

MaximProshin replied:
added the reference

alexsu52 commented on docs/optimization_guide/nncf/ptq/ptq_introduction.md:
```diff
@@ -10,17 +10,17 @@
-* :doc:`Basic quantization <basic_quantization_flow>` - the simplest quantization flow that allows to apply 8-bit integer quantization to the model.
-* :doc:`Quantization with accuracy control <quantization_w_accuracy_control>` - the most advanced quantization flow that allows to apply 8-bit quantization to the model with accuracy control.
+* :doc:`Basic quantization <basic_quantization_flow>` - the simplest quantization flow that allows applying 8-bit integer quantization to the model. A representative calibration dataset is only needed in this case.
+* :doc:`Quantization with accuracy control <quantization_w_accuracy_control>` - the most advanced quantization flow that allows applying 8-bit quantization to the model with accuracy control. Calibration and validation datasets, and accuracy metrics are needed in this case.
```
```suggestion
* :doc:`Quantization with accuracy control <quantization_w_accuracy_control>` - the most advanced quantization flow that allows applying 8-bit quantization to the model with accuracy control. Calibration and validation datasets, and validation function to calculate the accuracy metric are needed in this case.
```

MaximProshin replied:
updated

alexsu52 reviewed (commented):
A general comment from my side is that the structure is very complex. If the user wants to read about the basic quantization flow, he must make at least 4 clicks. I would recommend thinking about minimizing the number of hierarchy levels.
![image](https://github.com/openvinotoolkit/openvino/assets/3229971/219bbb0f-f93b-40d1-9c92-2fc5428b0de1)

MaximProshin commented:
> A general comment from my side is that the structure is very complex. If the user wants to read about the basic quantization flow, he must make at least 4 clicks. I would recommend thinking about minimizing the number of hierarchy levels. ![image](https://user-images.githubusercontent.com/3229971/238254001-219bbb0f-f93b-40d1-9c92-2fc5428b0de1.png)

With removal of POT in 2024, we will reduce one level so it will be Optimization=>PTQ=>Basic or AAQ. I don't know how to simplify it now (for me it looks ok). If you have concrete ideas how to restructure, please share it.

MaximProshin commented on docs/optimization_guide/nncf/ptq/basic_quantization_flow.md:
```diff
@@ -93,14 +92,14 @@ Once the dataset is ready and the model object is instantiated, you can apply 8-
 
 .. note:: The ``model`` is an instance of the ``torch.nn.Module`` class for PyTorch, ``onnx.ModelProto`` for ONNX, and ``openvino.runtime.Model`` for OpenVINO.
```
It means exactly what it says :) I guess it was added before we added code snippets so it was useful. Now it's covered there, so I removed it now.

tsavina commented:
> > A general comment from my side is that the structure is very complex. If the user wants to read about the basic quantization flow, he must make at least 4 clicks. I would recommend thinking about minimizing the number of hierarchy levels. ![image](https://user-images.githubusercontent.com/3229971/238254001-219bbb0f-f93b-40d1-9c92-2fc5428b0de1.png)
> 
> With removal of POT in 2024, we will reduce one level so it will be Optimization=>PTQ=>Basic or AAQ. I don't know how to simplify it now (for me it looks ok). If you have concrete ideas how to restructure, please share it.

We could reduce one level as we don't need to highlight POT as an option. Optimization=>Quantizing Models Post-training (contains PTQ with NNCF)=> Basic/Advanced/Deprecated POT

MaximProshin commented:
> > > A general comment from my side is that the structure is very complex. If the user wants to read about the basic quantization flow, he must make at least 4 clicks. I would recommend thinking about minimizing the number of hierarchy levels. ![image](https://user-images.githubusercontent.com/3229971/238254001-219bbb0f-f93b-40d1-9c92-2fc5428b0de1.png)
> > 
> > 
> > With removal of POT in 2024, we will reduce one level so it will be Optimization=>PTQ=>Basic or AAQ. I don't know how to simplify it now (for me it looks ok). If you have concrete ideas how to restructure, please share it.
> 
> We could reduce one level as we don't need to highlight POT as an option. Optimization=>Quantizing Models Post-training (contains PTQ with NNCF)=> Basic/Advanced/Deprecated POT

As we support POT in 2023, we still need to have the documentation for it

helena-intel commented on docs/optimization_guide/nncf/ptq/basic_quantization_flow.md:
```diff
@@ -150,12 +172,24 @@ Tune quantization parameters
        regex = '.*layer_.*'
        nncf.quantize(model, dataset, ignored_scope=nncf.IgnoredScope(patterns=regex))
 
+* ``target_device`` - defines the target device, the specificity of which will be taken into account during optimization. The following values are supported: ``ANY`` (default), ``CPU``, ``CPU_SPR``, ``GPU``, and ``VPU``.
```
It would be great to learn more about the effects of setting this parameter. I recently noticed a tiny accuracy degredation on SPR compared to ICL. Will specifying target_device=CPU_SPR fix this? Will that cause lower accuracy on ICL though? (Not asking for an answer, but that's what I wonder when I read this). Will this influence accuracy, performance or both?

MaximProshin replied:
CPU_SPR option improves model performance on SPR device due to removal of some FQs. It doesn't improve the accuracy, at least it doesn't solve the issue with bf16 enabled by default which I guess is the main reason of the accuracy degradation in your case.

helena-intel commented on docs/optimization_guide/nncf/ptq/ptq_introduction.md:
```diff
@@ -10,17 +10,17 @@
-* :doc:`Basic quantization <basic_quantization_flow>` - the simplest quantization flow that allows to apply 8-bit integer quantization to the model.
-* :doc:`Quantization with accuracy control <quantization_w_accuracy_control>` - the most advanced quantization flow that allows to apply 8-bit quantization to the model with accuracy control.
+* :doc:`Basic quantization <basic_quantization_flow>` - the simplest quantization flow that allows applying 8-bit integer quantization to the model. A representative calibration dataset is only needed in this case.
+* :doc:`Quantization with accuracy control <quantization_w_accuracy_control>` - the most advanced quantization flow that allows applying 8-bit quantization to the model with accuracy control. Calibration and validation datasets, and validation function to calculate the accuracy metric are needed in this case.
```
```suggestion
* :doc:`Quantization with accuracy control <quantization_w_accuracy_control>` - the most advanced quantization flow that allows applying 8-bit quantization to the model with accuracy control. Calibration and validation datasets, and a validation function to calculate the accuracy metric are needed in this case.
```

helena-intel commented on docs/optimization_guide/nncf/ptq/quantization_w_accuracy_control.md:
```diff
@@ -51,14 +51,33 @@ Now, you can run quantization with accuracy control. The following code snippet
 
 
-``max_drop`` defines the accuracy drop threshold. The quantization process stops when the degradation of accuracy metric on the validation dataset is less than the ``max_drop``.
+* ``max_drop`` defines the accuracy drop threshold. The quantization process stops when the degradation of accuracy metric on the validation dataset is less than the ``max_drop``. The default value is 0.01.
```
What happens if you set max_drop to an unreasonably low value that cannot be reached?

MaximProshin replied:
Just like POT it will stop the quantization and report an error that the max_drop value can't be reached. I've clarified it in the text.

helena-intel commented on docs/optimization_guide/nncf/ptq/quantization_w_accuracy_control.md:
```diff
@@ -51,14 +51,33 @@ Now, you can run quantization with accuracy control. The following code snippet
+      :fragment: [inference]
+
+
+``nncf.quantize_with_accuracy_control()`` API supports all the parameters from :doc:`Basic 8-bit quantization <basic_quantization_flow>` API and you can also use them to quantize a model with accuracy control and a custom configuration.
```
```suggestion
``nncf.quantize_with_accuracy_control()`` API supports all the parameters from :doc:`Basic 8-bit quantization <basic_quantization_flow>` API, to quantize a model with accuracy control and a custom configuration.
```
I had to read this sentence a few times. I think it's clearer without the "and you can also use them" part. Maybe someone has a better suggestion too.

MaximProshin replied:
updated as you suggested, I also think it's better

helena-intel commented on docs/optimization_guide/ptq_introduction.md:
```diff
@@ -29,7 +29,12 @@ To apply post-training methods in OpenVINO, you need:
-* :doc:`Post-training Quantization with POT <pot_introduction>` - works with models in OpenVINO Intermediate Representation (IR) only.
-* :doc:`Post-training Quantization with NNCF <nncf_ptq_introduction>` - cross-framework solution for model optimization that provides a new simple API for post-training quantization.
+* :doc:`Post-training Quantization with NNCF <nncf_ptq_introduction>` - cross-framework solution for model optimization that provides a simple API for post-training quantization.
+* :doc:`(Deprecated) Post-training Quantization with POT <pot_introduction>` - works with models in OpenVINO Intermediate Representation (IR) only.
```
Do we still want to mention POT here? It's deprecated. We should still have the "how to use POT" docs for existing users, but IMO this document can focus on NNCF, and doesn't have to say that we have "two workflows" anymore.

MaximProshin replied:
removed as I've changed the structure

MaximProshin commented on docs/optimization_guide/model_optimization_guide.md:
```diff
@@ -8,40 +8,29 @@
+- :doc:`Training-time Optimization <tmo_introduction>`, a suite of advanced methods for training-time model optimization within the DL framework, such as PyTorch and TensorFlow 2.x. It supports methods like Quantization-aware Training, Structured and Unstructured Pruning, etc. 
 
-- :doc:`Training-time Optimization <nncf_ptq_introduction>`, a suite of advanced methods for training-time model optimization within the DL framework, such as PyTorch and TensorFlow 2.x. It supports methods, like Quantization-aware Training and Filter Pruning. NNCF-optimized models can be inferred with OpenVINO using all the available workflows.
+.. note:: OpenVINO also supports optimized models (for example, quantized) from source frameworks such as PyTorch, TensorFlow, and ONNX (in Q/DQ format). No special steps are required in this case and optimized models can be converted to the OpenVINO Intermediate Representation format (IR) righ away.
```
done

MaximProshin commented on docs/optimization_guide/nncf/ptq/basic_quantization_flow.md:
```diff
@@ -29,25 +29,25 @@ Install all the packages required to instantiate the model object, for example,
 #############################
 
-At this step, create an instance of the ``nncf.Dataset`` class that represents the calibration dataset. The ``nncf.Dataset`` class can be a wrapper over the framework dataset object that is used for model training or validation. The class constructor receives the dataset object and the transformation function. For example, if you use PyTorch, you can pass an instance of the ``torch.utils.data.Dat ...
+At this step, create an instance of the ``nncf.Dataset`` class that represents the calibration dataset. The ``nncf.Dataset`` class can be a wrapper over the framework dataset object that is used for model training or validation. The class constructor receives the dataset object and the transformation function.
```
done

MaximProshin commented on docs/optimization_guide/model_optimization_guide.md:
```diff
@@ -8,40 +8,29 @@
+- :doc:`Training-time Optimization <tmo_introduction>`, a suite of advanced methods for training-time model optimization within the DL framework, such as PyTorch and TensorFlow 2.x. It supports methods like Quantization-aware Training, Structured and Unstructured Pruning, etc. 
 
-- :doc:`Training-time Optimization <nncf_ptq_introduction>`, a suite of advanced methods for training-time model optimization within the DL framework, such as PyTorch and TensorFlow 2.x. It supports methods, like Quantization-aware Training and Filter Pruning. NNCF-optimized models can be inferred with OpenVINO using all the available workflows.
+.. note:: OpenVINO also supports optimized models (for example, quantized) from source frameworks such as PyTorch, TensorFlow, and ONNX (in Q/DQ format). No special steps are required in this case and optimized models can be converted to the OpenVINO Intermediate Representation format (IR) righ away.
```
Yes, it's possible, but as far as I understand OVTF is deprecated or so. Unfortunately I wasn't able to find any documentation for ONNX OVEP, so I don't really know how to better refer to it.

MaximProshin commented:
> > > > > A general comment from my side is that the structure is very complex. If the user wants to read about the basic quantization flow, he must make at least 4 clicks. I would recommend thinking about minimizing the number of hierarchy levels. ![image](https://user-images.githubusercontent.com/3229971/238254001-219bbb0f-f93b-40d1-9c92-2fc5428b0de1.png)
> > > > 
> > > > 
> > > > With removal of POT in 2024, we will reduce one level so it will be Optimization=>PTQ=>Basic or AAQ. I don't know how to simplify it now (for me it looks ok). If you have concrete ideas how to restructure, please share it.
> > > 
> > > 
> > > We could reduce one level as we don't need to highlight POT as an option. Optimization=>Quantizing Models Post-training (contains PTQ with NNCF)=> Basic/Advanced/Deprecated POT
> > 
> > 
> > As we support POT in 2023, we still need to have the documentation for it
> 
> Maybe what we can do is to expose NNCF by default i.e. without separate subgroup for NNCF. And then, (deprecated) POT will be on the same level as Basic flow and advanced flow. In other words, move basic and advanced flow from NNCF upper level?

MaximProshin commented:
@yury-gorbachev , @alexsu52 , @tsavina  I addressed all your comments except for pictures which will be updated in a separate PR. So please check the updates.

MaximProshin commented:
> > @yury-gorbachev , @alexsu52 , @tsavina I addressed all your comments except for pictures which will be updated in a separate PR. So please check the updates.
> 
> Is documents link up to date also?

I think so. Is there something not working?

yury-gorbachev commented:
I like document after changes, it is simpler to read and one level less now.
One comment I have is - can we sync tabs in basic flow so that when I select framework - it applies to all code snippets? 
It is working here: http://openvino-doc.iotg.sclab.intel.com/optimization-docs/openvino_docs_OV_UG_Integrate_OV_with_your_application.html

yury-gorbachev commented:
> > I like document after changes, it is simpler to read and one level less now. One comment I have is - can we sync tabs in basic flow so that when I select framework - it applies to all code snippets? It is working here: http://openvino-doc.iotg.sclab.intel.com/optimization-docs/openvino_docs_OV_UG_Integrate_OV_with_your_application.html
> 
> @yury-gorbachev Done. Please take a look.

Looks great!

alexsu52 commented on docs/optimization_guide/nncf/ptq/basic_quantization_flow.md:
```diff
@@ -57,15 +57,20 @@ The transformation function is a function that takes a sample from the dataset a
        :language: python
        :fragment: [dataset]
 
+If there is no framework dataset object, you can create your own entity that implements the ``Iterable`` interface in Python, for example the list of images, and returns data samples feasible for inference. In this case, a transformation function is not required.
```
Hi Helena, I absolutely agree with you that `for example the list of images` is confusing.The main meaning of this sentence that nncf.Dataset supports for any iterable Python object over a dataset and user can implement or use any entity that implements Python iterable interface . If custom or framework dataset returns data samples feasible for inference then a transformation function is not required.

I would recommend to reformulate this sentence:
```
If there is no framework dataset object, you can create your own entity that implements the ``Iterable`` interface in Python. If dataset object returns data samples feasible for inference then a transformation function is not required.
```
I don't seen any reason to add example here because example will include the following code:
```
model_inputs = [...] # model(model_inputs[0])

calibration_dataset = nncf.Dataset(model_inputs)
```
Not sure if this will bring any additional information. If you think otherwise, we can do it.

MaximProshin replied:
@helena-intel , if you have a suggestion how to better formulate it, feel free to share it here.

helena-intel replied:
I think an example of such an iterable would be very useful, also because the PTQ OpenVINO example use an Ultralytics or PyTorch dataloader, and if you're not familiar with them, it may not be obvious what they return, how they handle batching, etc. I'll try to create the most basic NNCF PTQ example today. We can see how to incorporate that in either docs or examples, but that can be in a future PR.

alexsu52 commented on docs/optimization_guide/nncf/ptq/code/ptq_torch.py:
```diff
@@ -19,3 +19,18 @@ def transform_fn(data_item):
+import openvino.runtime as ov
+from openvino.tools.mo import convert_model
+
+input_fp32 = ... # FP32 model input
+
+# convert PyTorch model to OpenVINO model
+ov_quantized_model = convert_model(quantized_model, example_input=input_fp32)
```
```suggestion
input_fp32 = ... # FP32 model input

# export PyTorch model to ONNX model
onnx_model_path = "model.onnx"
torch.onnx.export(quantized_model, input_fp32, onnx_model_path)

# convert ONNX model to OpenVINO model
ov_quantized_model = convert_model(onnx_model_path)
```

Updated the script according the changes in the OV 2023.0.0 cc' @slyalin @yury-gorbachev

alexsu52 commented on docs/optimization_guide/nncf/ptq/basic_quantization_flow.md:
```diff
@@ -29,78 +29,118 @@ Install all the packages required to instantiate the model object, for example,
-    .. doxygensnippet:: docs/optimization_guide/nncf/ptq/code/ptq_onnx.py
-       :language: python
-       :fragment: [quantization]
+After that the model can be converted into the OpenVINO Intermediate Representation (IR) if needed, compiled and run with OpenVINO:
```
> It would be good to also explain how to save the quantized model to disk. Like `ov.serialize(quantized_model, "quantized_model.xml")`

It looks like this is common OpenVINO API.

helena-intel replied:
It is! Just like the OpenVINO inference code above. Not every person who learns how to quantize a model with NNCF will be familiar with how to serialize models in OpenVINO, so it's good to document it.

alexsu52 commented on docs/optimization_guide/nncf/ptq/code/ptq_tensorflow.py:
```diff
@@ -19,3 +19,17 @@ def transform_fn(data_item):
+# compile the model to transform quantized operations to int8
+model_int8 = ov.compile_model(ov_quantized_model)
+
+input_fp32 = ... # FP32 model input
+res = model_int8(input_fp32)
+#! [inference]
```
```suggestion
input_fp32 = ... # FP32 model input
res = model_int8(input_fp32)

# save the model
ov.serialize(ov_quantized_model, "quantized_model.xml")
#! [inference]

```

alexsu52 commented on docs/optimization_guide/nncf/ptq/code/ptq_torch.py:
```diff
@@ -19,3 +19,22 @@ def transform_fn(data_item):
+# compile the model to transform quantized operations to int8
+model_int8 = ov.compile_model(ov_quantized_model)
+
+res = model_int8(input_fp32)
+#! [inference]
```
```suggestion
res = model_int8(input_fp32)

# save the model
ov.serialize(ov_quantized_model, "quantized_model.xml")
#! [inference]

```

alexsu52 commented on docs/optimization_guide/nncf/ptq/code/ptq_openvino.py:
```diff
@@ -19,3 +19,13 @@ def transform_fn(data_item):
+model_int8 = ov.compile_model(quantized_model)
+
+input_fp32 = ... # FP32 model input
+res = model_int8(input_fp32)
+#! [inference]
```
```suggestion
res = model_int8(input_fp32)

# save the model
ov.serialize(quantized_model, "quantized_model.xml")
#! [inference]

```

alexsu52 commented on docs/optimization_guide/nncf/ptq/code/ptq_onnx.py:
```diff
@@ -20,3 +20,17 @@ def transform_fn(data_item):
+# compile the model to transform quantized operations to int8
+model_int8 = ov.compile_model(ov_quantized_model)
+
+input_fp32 = ... # FP32 model input
+res = model_int8(input_fp32)
+#! [inference]
```
```suggestion
input_fp32 = ... # FP32 model input
res = model_int8(input_fp32)

# save the model
ov.serialize(ov_quantized_model, "quantized_model.xml")
#! [inference]

```

alexsu52 commented on docs/optimization_guide/nncf/ptq/code/ptq_aa_openvino.py:
```diff
@@ -45,5 +45,16 @@ def validate(model: openvino.runtime.CompiledModel,
+# compile the model to transform quantized operations to int8
+model_int8 = ov.compile_model(quantized_model)
+
+input_fp32 = ... # FP32 model input
+res = model_int8(input_fp32)
+#! [inference]
```
```suggestion
input_fp32 = ... # FP32 model input
res = model_int8(input_fp32)

# save the model
ov.serialize(quantized_model, "quantized_model.xml")
#! [inference]

```

MaximProshin commented:
@tsavina , everything should be ok now, let's merge it!
